# Optimization

## Homework 4

_October 6th, 2020_

### Problem 1

(Beck Exercise 5.2) Consider the Freudenstein and Roth test function

$$f(x) = f_1(x)^2 + f_2(x)^2, \ \ x \in \mathbb R^2$$

where
$$f_1(x) = -13 + x_1 + ((5-x_2)x_2-2)x_2$$

$$f_2(x) = -29 + x_1 + ((1+x_2)x_2-14)x_2$$

**(i)** Show that the function $f$ has three stationary points. Find them and prove that one is a global minimizer, one is a strict local minimum, and the third is a saddle point.

(see HW4.pdf for this work. Some of the intermediate calculations required are below)

In [1]:
import numpy as np
from numpy.linalg import norm
from numpy.linalg import cholesky
from numpy.linalg import solve

In [2]:
from IPython.display import HTML
HTML('''<script>
code_show_err=false; 
function code_toggle_err() {
 if (code_show_err){
 $('div.output_stderr').hide();
 } else {
 $('div.output_stderr').show();
 }
 code_show_err = !code_show_err
} 
$( document ).ready(code_toggle_err);
</script>
To toggle on/off output_stderr, click <a href="javascript:code_toggle_err()">here</a>.''')

In [3]:
# objective function
f1 = lambda x: -13 + x[0] + ((5 - x[1])*x[1] - 2)*x[1]
f2 = lambda x: -29 + x[0] + ((1 + x[1])*x[1] - 14)*x[1]

fvec = lambda x: np.array([f1(x),f2(x)])

f = lambda x: fvec(x).T@fvec(x) 

In [4]:
# gradient
g = lambda x: np.array([-84 + 4*x[0] + 12*x[1]**2 - 32*x[1],
                        864 - 32*x[0] + 24*x[1] + 24*x[0]*x[1] - 240*x[1]**2 + 8*x[1]**3 - 40*x[1]**4 + 12*x[1]**5])

In [5]:
# global minimum
x1 = np.array([5,4])
print("gradient at x1: ", g(x1), "function value at x2: ", f(x1))

gradient at x1:  [0 0] function value at x2:  0


In [6]:
# local minimum
x2 = np.array([(1/3)*(53-4*np.sqrt(22)), (1/3)*(2-np.sqrt(22))])
print("gradient at x2: ", g(x2), "function value at x2: ", f(x2))

gradient at x2:  [-3.55271368e-15  1.33226763e-14] function value at x2:  48.98425367924004


In [7]:
# saddle point
x3 = np.array([(1/3)*(53+4*np.sqrt(22)), (1/3)*(2+np.sqrt(22))])
print("gradient at x3: ", g(x3), "function value at x3: ", f(x3))

gradient at x3:  [0. 0.] function value at x3:  819.01025935231


In [8]:
# hessian
h = lambda x: np.array([[4, 24*x[1]-32],
                        [24*x[1]-32 ,24 + 24*x[0] - 480*x[1] + 24*x[1]**2 - 160*x[1]**3 + 60*x[1]**4]])

In [9]:
h(x1)

array([[   4,   64],
       [  64, 3728]])

In [10]:
h(x2)

array([[  4.        , -53.52332608],
       [-53.52332608, 901.88775184]])

In [11]:
h(x3)

array([[   4.        ,   21.52332608],
       [  21.52332608, -643.51738147]])

**(ii)** Employ the following three methods on the problem of minimizing f. 

### Gradient Method with backtracking
**(a)** Gradient Method with backtracking and parameters $(s, \alpha, \beta) = (1, 0.5, 0.5)$.

In [12]:
def gradient_method_backtracking(f,g,x0,s,alpha,beta,epsilon):
    """
    Gradient method with backtracking stepsize rule
    INPUT
    =======================================
    f ......... objective function
    g ......... gradient of the objective function
    x0......... initial point
    s ......... initial choice of stepsize
    alpha ..... tolerance parameter for the stepsize selection
    beta ...... the constant in which the stepsize is multiplied 
                at each backtracking step (0<beta<1)
    epsilon ... tolerance parameter for stopping rule
    OUTPUT
    =======================================
    x ......... optimal solution (up to a tolerance) 
                of min f(x)
    fun_val ... optimal function value
    """
    x=x0
    grad=g(x)
    fun_val=f(x)
    iter=0
    while (norm(grad)>epsilon):
        iter=iter+1
        t=s
        while (fun_val-f(x-t*grad)<alpha*t*norm(grad)**2):
            t=beta*t
        x=x-t*grad
        fun_val=f(x)
        grad=g(x)
    print('iter_number = '+ str(iter) + ' norm_grad = ' + str(norm(grad)) + ' fun_val = ' + str(fun_val))
    return x,fun_val

In [13]:
# parameters
s = 1
alpha = 0.5 
beta = 0.5
epsilon = 10e-5 

In [14]:
# starting point
x0 = np.array([-50,7])
gradient_method_backtracking(f,g,x0,s,alpha,beta,epsilon)

C:\Users\saeth\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: overflow encountered in long_scalars
  
C:\Users\saeth\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: RuntimeWarning: overflow encountered in long_scalars
  This is separate from the ipykernel package so we can avoid doing imports until


iter_number = 1965 norm_grad = 9.980272066889434e-05 fun_val = 1.689658077800942e-09


(array([4.99996587, 4.00000058]), 1.689658077800942e-09)

#### Starting at $(-50,7)$ it took 1965 iterations to converge.

In [15]:
# starting point
x0 = np.array([20,7])
gradient_method_backtracking(f,g,x0,s,alpha,beta,epsilon)

C:\Users\saeth\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: overflow encountered in long_scalars
  
C:\Users\saeth\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: RuntimeWarning: overflow encountered in long_scalars
  This is separate from the ipykernel package so we can avoid doing imports until


iter_number = 2155 norm_grad = 9.930646621307876e-05 fun_val = 1.6399482633305737e-09


(array([5.00003362, 3.99999943]), 1.6399482633305737e-09)

#### Starting at $(20,7)$ it took 2155 iterations to converge.

In [16]:
# starting point
x0 = np.array([20,-18])
gradient_method_backtracking(f,g,x0,s,alpha,beta,epsilon)

C:\Users\saeth\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: overflow encountered in long_scalars
  
C:\Users\saeth\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: RuntimeWarning: overflow encountered in long_scalars
  This is separate from the ipykernel package so we can avoid doing imports until


iter_number = 2117 norm_grad = 9.717228016084474e-05 fun_val = 48.98425368488135


(array([11.41289603, -0.89679829]), 48.98425368488135)

#### Starting at $(20,-18)$ it took 2117 iterations to converge.

In [17]:
# starting point
x0 = np.array([5,-10])
gradient_method_backtracking(f,g,x0,s,alpha,beta,epsilon)

C:\Users\saeth\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: overflow encountered in long_scalars
  
C:\Users\saeth\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: RuntimeWarning: overflow encountered in long_scalars
  This is separate from the ipykernel package so we can avoid doing imports until


iter_number = 1832 norm_grad = 9.553119633971772e-05 fun_val = 1.5451971987425595e-09


(array([5.00003264, 3.99999944]), 1.5451971987425595e-09)

#### Starting at $(5,-10)$ it took 1832 iterations to converge.

### Hybrid Newton's Method
**(b)** Hybrid Newton's Method with $(\alpha, \beta) = (0.5,0.5)$

In [18]:
def hybrid_newton(f,g,x0,h,alpha,beta,epsilon):
    """
    Hybrid Newton's method
    INPUT
    =======================================
    f ......... objective function
    g ......... gradient of the objective function
    h ......... hessian of the objective function
    x0......... initial point
    alpha ..... tolerance parameter for the stepsize selection
    beta ...... the constant in which the stepsize is multiplied 
                at each backtracking step (0<beta<1)
    epsilon ... tolerance parameter for stopping rule
    OUTPUT
    =======================================
    x ......... optimal solution (up to a tolerance) 
                of min f(x)
    fun_val ... optimal function value
    """
    x=x0
    grad=g(x)
    hval = h(x)
    
    try:
        L = cholesky(hval)
        d = solve(L.T, solve(L, grad))
    except:
        d = grad
    
    fun_val=f(x)
    iter=0
    while (norm(grad)>epsilon) & (iter<10000):
        iter=iter+1
        t=1
        while (f(x-t*d) > f(x) - alpha*t*grad.T@d):
            t=beta*t
        x=x-t*d
        fun_val=f(x)
        grad=g(x)
        hval = h(x)
        try:
            L = cholesky(hval)
            d = solve(L.T, solve(L, grad))
        except:
            d = grad
            
    if(iter == 10000):
        print("method did not converge")
        
    print('iter_number = '+ str(iter) + ' norm_grad = ' + str(norm(grad)) + ' fun_val = ' + str(fun_val))
    
    return x,fun_val

In [19]:
x0 = np.array([-50,7])
hybrid_newton(f,g,x0,h,alpha,beta,epsilon)

iter_number = 8 norm_grad = 4.403801190986987e-09 fun_val = 3.218858347296457e-21


(array([5., 4.]), 3.218858347296457e-21)

#### Starting at $(-50,7)$ it took 8 iterations to converge.

In [20]:
x0 = np.array([20,7])
hybrid_newton(f,g,x0,h,alpha,beta,epsilon)

iter_number = 8 norm_grad = 6.753950355311321e-09 fun_val = 7.573184445518013e-21


(array([5., 4.]), 7.573184445518013e-21)

#### Starting at $(20,7)$ it took 8 iterations to converge.

In [21]:
x0 = np.array([20,-18])
hybrid_newton(f,g,x0,h,alpha,beta,epsilon)

iter_number = 15 norm_grad = 5.948523474557842e-05 fun_val = 48.98425367924293


(array([11.41277747, -0.89680541]), 48.98425367924293)

#### Starting at $(20,-18)$ it took 15 iterations to converge.

In [22]:
x0 = np.array([5,-10])
hybrid_newton(f,g,x0,h,alpha,beta,epsilon)

iter_number = 13 norm_grad = 4.285820184238037e-08 fun_val = 48.98425367924004


(array([11.41277899, -0.89680525]), 48.98425367924004)

#### Starting at $(5,-10)$ it took 13 iterations to converge.

### Damped Gauss-Newton Method
**(c)**  Damped Gauss-Newton Method with backtracking line search strategy with $(s,\alpha, \beta) = (1,0.5,0.5)$ 

The Gauss-Newton Method is essentially  a scaled gradient method with the following positive definite-scaling matrix

$$D_k = \frac{1}{2}(J(x_k)^TJ(x_k))^{-1}$$

In [29]:
def scaled_gradient_method_backtracking(J,f,g,x0,s,alpha,beta,epsilon):
    """
    Scaled gradient method with backtracking stepsize rule
    INPUT
    =======================================
    J ......... scaling matrix
    f ......... objective function
    g ......... gradient of the objective function
    x0......... initial point
    s ......... initial choice of stepsize
    alpha ..... tolerance parameter for the stepsize selection
    beta ...... the constant in which the stepsize is multiplied 
                at each backtracking step (0<beta<1)
    epsilon ... tolerance parameter for stopping rule
    OUTPUT
    =======================================
    x ......... optimal solution (up to a tolerance) 
                of min f(x)
    fun_val ... optimal function value
    """
    x=x0
    grad=g(x)
    fun_val=f(x)
    D = (1/2)*np.linalg.inv(J(x).T@J(x))
    iter=0
    while ((norm(grad)>epsilon)&(iter<10000)):
        iter=iter+1
        t=s
        while (fun_val-f(x-t*D@grad)<alpha*t*grad@D@grad):
            t=beta*t
        x=x-t*D@grad
        fun_val=f(x)
        grad=g(x)
        D = (1/2)*np.linalg.inv(J(x).T@J(x))
    if (iter==10000):
        print('did not converge')
    print('iter_number = '+ str(iter) + ' norm_grad = ' + str(norm(grad)) + ' fun_val = ' + str(fun_val))
    return x,fun_val

In [30]:
# jacobian
J = lambda x: np.array([[1, 10 - 27*x[1] + 10*x[1]**2 - x[1]**3],
                        [1, -14 - 13*x[1] + 2*x[1]**2 + x[1]**3]])

In [31]:
x0 = np.array([-50,7])
scaled_gradient_method_backtracking(J,f,g,x0,s,alpha,beta,epsilon)

iter_number = 21 norm_grad = 2.003819082630959e-05 fun_val = 4.47489664292633e-12


(array([5.00000175, 3.99999996]), 4.47489664292633e-12)

In [33]:
x0 = np.array([20,7])
scaled_gradient_method_backtracking(J,f,g,x0,s,alpha,beta,epsilon)

iter_number = 22 norm_grad = 8.265639256302167e-05 fun_val = 1.00716430311037e-12


(array([5.00000025, 4.00000002]), 1.00716430311037e-12)

In [35]:
x0 = np.array([20,18])
scaled_gradient_method_backtracking(J,f,g,x0,s,alpha,beta,epsilon)

iter_number = 63 norm_grad = 7.13070672683346e-05 fun_val = 3.5420750664235985e-12


(array([5.0000014 , 3.99999996]), 3.5420750664235985e-12)

In [36]:
x0 = np.array([5,-10])
scaled_gradient_method_backtracking(J,f,g,x0,s,alpha,beta,epsilon)

iter_number = 879 norm_grad = 9.82011484157245e-05 fun_val = 48.98425367938374


(array([11.41276065, -0.89680645]), 48.98425367938374)